### PR TITLE
8345299: C2: some nodes can still have incorrect control after do_range_check()

### PR DESCRIPTION
**Pull request by rwestrel**
8339733 fixed controls for updated pre/main limits during
do_range_check(). However, it missed one issue:

Control for the new limits is computed in `new_limit_ctrl` for both
pre and main loops. `new_limit_ctrl` is currently initialized from the
pre limit control but it also needs to take the main loop limit
control into account as sometimes, the main loop limit control is
below the pre limit and pre loop entry control.

8339733 also introduced a couple bugs. Control of the `Bool`/`Cmp`
nodes are updated for the pre and main loop to `new_limit_ctrl`. But
that's incorrect because `new_limit_ctrl` may be above the pre loop
while the `Bool`/`Cmp` for the pre loop are in the loop (because they
depend on the loop iv) and for the main loop are after the pre loop
(because they depend on the iv out of the pre loop). I fixed this for
the pre loop by setting control for the `Bool`/`Cmp` to be as late as
possible. For the main loop, no change appears to be required as
control computed by c2 is already late enough. I've added an assert
instead.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8345299](https://bugs.openjdk.org/browse/JDK-8345299): C2: some nodes can still have incorrect control after do_range_check() (**Bug** - P4)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22485/head:pull/22485` \
`$ git checkout pull/22485`

Update a local copy of the PR: \
`$ git checkout pull/22485` \
`$ git pull https://git.openjdk.org/jdk.git pull/22485/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22485`

View PR using the GUI difftool: \
`$ git pr show -t 22485`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22485.diff">https://git.openjdk.org/jdk/pull/22485.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22485#issuecomment-2511625644)
</details>

### CODE REVIEW

**Review by chhagedorn — APPROVED** [review]
Looks good to me. Do you have a regression test where this leads to an actual failure?

**Comment by rwestrel** [comment]
> Looks good to me. Do you have a regression test where this leads to an actual failure?

I don't. This seems unlikely to cause issues with the current code. I ran into this while working on JDK-8275202 where new code I've added is failing.

**Comment by chhagedorn** [comment]
I see, thanks for background!

File: src/hotspot/share/opto/loopTransform.cpp [hunk 1]
```diff
@@ -2778,8 +2779,8 @@ void PhaseIdealLoop::do_range_check(IdealLoopTree *loop, Node_List &old_new) {
   // new pre_limit can push Bool/Cmp/Opaque nodes down (when one of the eliminated condition has parameters that are not
   // loop invariant in the pre loop.
   set_ctrl(pre_opaq, new_limit_ctrl);
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopTransform.cpp:2781]: Can you update this comment to explain different control settings as you did in PR's description.

File: src/hotspot/share/opto/loopTransform.cpp [hunk 2]
```diff
@@ -2820,10 +2821,9 @@ void PhaseIdealLoop::do_range_check(IdealLoopTree *loop, Node_List &old_new) {
   assert(opqzm->outcnt() == 1, "cannot hack shared node");
   _igvn.replace_input_of(opqzm, 1, main_limit);
   // new main_limit can push Bool/Cmp nodes down (when one of the eliminated condition has parameters that are not loop
-  // invariant in the pre loop.
+  // invariant in the pre loop).
   set_ctrl(opqzm, new_limit_ctrl);
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopTransform.cpp:2825]: Update this comment too.

**Review by vnkozlov — COMMENTED** [review]
Few comments.

File: src/hotspot/share/opto/loopTransform.cpp (hunk 2 above)
Comment (rwestrel) [reply src/hotspot/share/opto/loopTransform.cpp]: Both done in new commit. Do the comments look ok to you?

**Review by vnkozlov — APPROVED** [review]
Good.

**Review by chhagedorn — APPROVED** [review]

**Comment by rwestrel** [comment]
@vnkozlov @chhagedorn thanks for the review

**Pull request closed without merging**